Product Collection: Add author filter in sidebar settings

imanish003 opened the pull request:
This PR adds an author filter to the ProductCollection block
<img width="272" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/e63257e6-a752-4a09-9d08-c062edfaf7e1">


- A new file, `author-control.tsx`, has been created inside the `inspector-controls` directory of the `product-collection` block. This file contains the React component for the author filter, which fetches the list of authors from the WordPress database using `useEntityRecords`. The filter provides an interface for selecting authors whose products should be displayed in the ProductCollection block.

- The author filter is added to the main component of the `inspector-controls` in `index.tsx`.

- In `ProductCollection.php`, the 'author' parameter is now included in the product query. This allows products from a specific author to be displayed in the ProductCollection block, depending on the user's selection in the Gutenberg editor.

This addition enhances the functionality of the ProductCollection block by allowing site administrators to create more customized and targeted displays of products.

Fixes #9366

### Testing

##### Simplest path

1. Add a “Product Collection” block to your page.
2. Go to the “Filters” within the Inspector Controls and add “Authors”.
3. Start typing to select authors. If you don't have products with different authors then please add some for testing.
4. As you narrow down your selection, verify that the preview is updated accurately. Only products by selected authors should be visible. 
5. Select 'Publish' to apply the changes.
6. Ensure the front-end is correct.

##### Complex path

1. Repeat steps 1–4 above.
2. Add additional filters (such as “On sale”).
3. With these additional filters applied, ensure that the products continue to be filtered accurately based on the selected authors.

Here is a quick demo:

https://github.com/woocommerce/woocommerce-blocks/assets/16707866/c213648c-79df-4134-bc61-7d33231d8bdb

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

### Changelog

> Product Collection: Add author filter in sidebar settings

imanish003 committed:
Add author filter to ProductCollection block

This commit adds an author filter to the ProductCollection block in WooCommerce Blocks.

- A new file, `author-control.tsx`, has been created inside the `inspector-controls` directory of the `product-collection` block. This file contains the React component for the author filter, which fetches the list of authors from the WordPress database using `useEntityRecords`. The filter provides an interface for selecting authors whose products should be displayed in the ProductCollection block.

- The author filter is added to the main component of the `inspector-controls` in `index.tsx`.

- In `ProductCollection.php`, the 'author' parameter is now included in the product query. This allows products from a specific author to be displayed in the ProductCollection block, depending on the user's selection in the Gutenberg editor.

This addition enhances the functionality of the ProductCollection block by allowing site administrators to create more customized and targeted displays of products.

imanish003 committed:
Update error handling

kmanijak commented on assets/js/blocks/product-collection/inspector-controls/author-control.tsx:
```diff
@@ -0,0 +1,158 @@
+	name: string;
+}
+
+interface AuthorsMapping {
+	mapById: Map< number, Author >;
+	mapByName: Map< string, Author >;
+	names: string[];
+}
```
This interface is used only to be extended by `AuthorsInfo`. In that case, I think it could be moved to the `AuthorsInfo` definition without extending unless there's some reason behind that split?

imanish003 replied:
I appreciate your input. Upon further review, I realized that I was utilizing `AuthorsMapping` in two different areas. However, due to some recent code changes, I am no longer using it elsewhere. Thank you for bringing this to my attention. I will merge it into `AuthorsInfo`. 🙌🏻

kmanijak reviewed (changes requested):
Great work on that!

I found an issue on the frontend though presented in this video:

https://github.com/woocommerce/woocommerce-blocks/assets/20098064/86b2d031-2d67-4d91-b732-f135bdf6c20d

For some reason, pagination is broken, every page is showing the same content. Issue doesn't occur on `trunk` so seems it may be related to this change.

imanish003 committed:
Refactor 'Author' interfaces and fix offset calculation

This commit refactors the 'Author' interfaces in the 'author-control.tsx' file. The interface 'AuthorsMapping' has been removed and its properties have been merged into 'AuthorsInfo' for simplicity and maintainability.

In the PHP file 'ProductCollection.php', the offset calculation logic has been improved to handle pagination correctly. Now, the offset is calculated as '(per_page * (page - 1)) + offset', which correctly accounts for both the page number and the offset when determining the starting point for the product query.

imanish003 commented:
> For some reason, pagination is broken, every page is showing the same content. Issue doesn't occur on trunk so seems it may be related to this change.

@kmanijak Thank you so much for bringing this to my attention. The issue is happening for me on trunk as well, and I believe it's expected behavior because of wrong offset calculation. I have addressed and fixed it in [2be10c9](https://github.com/woocommerce/woocommerce-blocks/pull/9736/commits/2be10c9387449e656d9d12d693af216414b0a520). Could you please take another look?

imanish003 committed:
Merge branch 'trunk' into 9366-product-collection-filters-authors

imanish003 committed:
Merge branch 'trunk' into 9366-product-collection-filters-authors

kmanijak reviewed (approved):
It works as expected! And thanks for fixing the offset issue 💪
